src/bin/tectonic.rs

```diff
@@ -19,49 +20,91 @@ use tectonic::status::{ChatterLevel, StatusBackend};
 
 use tectonic::{ctry, errmsg, tt_error, tt_error_styled, tt_note};
 
-fn inner(
-    args: ArgMatches,
-    config: PersistentConfig,
-    status: &mut TermcolorStatusBackend,
-) -> Result<()> {
+#[derive(Debug, StructOpt)]
+#[structopt(name = "Tectonic", about = "Process a (La)TeX document")]
+struct Opt {
+    /// The file to process, or "-" to process the standard input stream"
+    #[structopt(name = "INPUT")]
+    input: String,
+    /// The name of the "format" file used to initialize the TeX engine
+    #[structopt(long, short, name = "PATH", default_value = "latex")]
+    format: String,
+    /// Use this Zip-format bundle file to find resource files instead of the default
+    #[structopt(
+        takes_value(true),
+        parse(from_os_str),
+        long,
+        short,
+        name = "zip_file_path"
+    )]
+    bundle: Option<PathBuf>,
+    /// Use this URL find resource files instead of the default
+    #[structopt(takes_value(true), long, short, name = "url")]
+    // TODO add URL validation
+    web_bundle: Option<String>,
+    /// How much chatter to print when running
+    #[structopt(long = "chatter", short, name = "LEVEL", default_value = "default", possible_values(&["default", "minimal"]))]
+    chatter_level: String,
+    /// Use only resource files cached locally
+    #[structopt(short = "C")]
+    only_cached: bool,
+    /// The kind of output to generate
+    #[structopt(long, name = "format", default_value = "pdf", possible_values(&["pdf", "html", "xdv", "aux", "format"]))]
+    outfmt: String,
+    /// Write Makefile-format rules expressing the dependencies of this run to <DEST_PATH>
+    #[structopt(long, name = "DEST_PATH")]
+    makefile_rules: Option<PathBuf>,
+    /// Which engines to run
+    #[structopt(long, default_value = "default", possible_values(&["default", "tex", "bibtex_first"]))]
+    pass: String,
+    /// Rerun the TeX engine exactly this many times after the first
+    #[structopt(name = "count", long = "reruns", short = "r")]
+    reruns: Option<usize>,
+    /// Keep the intermediate files generated during processing
+    #[structopt(short, long)]
+    keep_intermediates: bool,
+    /// Keep the log files generated during processing
+    #[structopt(long)]
+    keep_logs: bool,
+    /// Generate SyncTeX data
+    #[structopt(long)]
+    synctex: bool,
+    /// Tell the engine that no file at <HIDE_PATH> exists, if it tries to read it
+    #[structopt(long, name = "hide_path")]
+    hide: Option<Vec<PathBuf>>,
+    /// Print the engine's chatter during processing
+    #[structopt(long = "print", short)]
+    print_stdout: bool,
+    /// The directory in which to place output files [default: the directory containing INPUT]
+    #[structopt(name = "OUTDIR", short, long, parse(from_os_str))]
+    outdir: Option<PathBuf>,
+}
+fn inner(args: Opt, config: PersistentConfig, status: &mut TermcolorStatusBackend) -> Result<()> {
```

Comment:
Random fun idea, for certain values of "fun": can we implement `TryFrom<CliOptions>` for `ProcessingSession`?

Reply:
@pkgw I have a vague idea of how to do this so perhaps I could try in another PR, or if you get to it I can review. I am stuck on how to share `CliOptions` from `bin/tectonic.rs` to the `driver.rs` file. ([This](https://github.com/efx/tectonic/compare/topic-structopt...efx:topic-structopt-tryfrom?expand=1) is as far as I got)

Reply:
Definitely no worries ­I don't even know if my idea even works, and it wouldn't actually provide any additional functionality. (I *think* you can impl the trait inside `bin/tectonic.rs` since the type parameter would be a type local to the binary crate, but I didn't check whether that's possible.) So, yeah, this should be punted.